### ISSUE TITLE
[Bug] Pro MaxMindDetailsAddon references removed GeoIP class — fatal on Advanced Whois AJAX

### ISSUE THREAD

**Issue by parhumm** [slimstat/wp-slimstat#182]
## Description

The Pro plugin's `MaxMindDetailsAddon` (Advanced Whois feature) references `SlimStat\Services\GeoIP` class and suppress-loads `/wp-slimstat/src/Services/GeoIP.php`, but the free plugin no longer ships that file or class.

The geolocation stack was refactored to:
- `SlimStat\Services\Geolocation\GeolocationService`
- `SlimStat\Services\Geolocation\GeolocationFactory`
- `SlimStat\Services\Geolocation\Provider\*` (DbIpProvider, MaxmindGeoIPProvider, CloudflareGeolocationProvider)
- `SlimStat\Services\Geolocation\AbstractGeoIPProvider`

## Affected Code

`wp-slimstat-pro/src/Addon/Addons/MaxMindDetailsAddon.php`:

- **Line 86**: `!\SlimStat\Services\GeoIP::database_exists()` — class does not exist
- **Line 91**: `@include_once(WP_PLUGIN_DIR . '/wp-slimstat/src/Services/GeoIP.php')` — file does not exist
- **Line 92**: `GeoIP::loader($ip)` — class does not exist

## Impact

When a user clicks the Advanced Whois/IP Details link in the SlimStat admin reports, the AJAX handler will fatal error because the included file and class no longer exist. The `@include_once` suppresses the include warning but `GeoIP::loader()` on line 92 will cause a fatal "Class not found" error.

## Steps to Reproduce

1. Activate wp-slimstat + wp-slimstat-pro
2. View SlimStat reports
3. Click on an IP address to view Whois/geolocation details
4. Expected: IP details popup with geolocation info
5. Actual: Fatal error (Class 'GeoIP' not found)

## Suggested Fix

Update `MaxMindDetailsAddon.php` to use the new geolocation API:
- Replace `GeoIP::database_exists()` with a check via `GeolocationFactory` or the `geolocation_provider` setting
- Replace `GeoIP::loader($ip)` with `GeolocationService::locate($ip)` using the configured provider

## Discovery Context

Found during PR #152 (Cloudflare IP geolocation fix) QA verification. This is a pre-existing issue not introduced by that PR.

**Comment by parhumm** [comment]
## Validation Result: Confirmed Bug

**Verdict**: VALID_BUG | **Confidence**: HIGH | **Severity**: High

### Summary

Confirmed that `MaxMindDetailsAddon` references the removed `SlimStat\Services\GeoIP` class in 4 locations. The free plugin refactored geolocation into `SlimStat\Services\Geolocation\*` namespace (commit `7e5ceac8`) but the Pro addon was not updated, causing a fatal error when Advanced Whois is used.

### Root Cause

**Fault Location**: `wp-slimstat-pro/src/Addon/Addons/MaxMindDetailsAddon.php:86-92,159`
**Mechanism**: Geolocation API refactored in free plugin; Pro addon still references deleted `GeoIP` class and file.

**Causal Chain**:
User clicks IP in reports → `ip2locationIframeContent()` AJAX handler → `@include_once` silently fails (file deleted) → `GeoIP::loader($ip)` fatal (class not found)

**5 Whys**:
1. Why fatal error? → `GeoIP::loader($ip)` on non-existent class
2. Why is class missing? → `@include_once` silently fails since `GeoIP.php` deleted
3. Why was it deleted? → Geolocation refactored to `Services/Geolocation/` namespace
4. Why wasn't Pro updated? → Refactor scope was free plugin only ← ROOT CAUSE

### Reproduction Steps

**Preconditions**: WP + wp-slimstat (v5.4+) + wp-slimstat-pro, Advanced Whois enabled

1. Navigate to WP Admin → SlimStat reports
2. Click any visitor IP address for Advanced Whois details
3. AJAX request fires to `admin-ajax.php?action=slimstat_ip2location_iframe_content`

**Expected**: IP geolocation details popup with country, city, coordinates, activity
**Actual**: Fatal PHP error — Class 'SlimStat\Services\GeoIP' not found

### Code References

| File | Lines | Description |
|------|-------|-------------|
| `wp-slimstat-pro/src/Addon/Addons/MaxMindDetailsAddon.php` | 6 | `use SlimStat\Services\GeoIP` — class does not exist |
| `wp-slimstat-pro/src/Addon/Addons/MaxMindDetailsAddon.php` | 86 | `GeoIP::database_exists()` — no such method in new API |
| `wp-slimstat-pro/src/Addon/Addons/MaxMindDetailsAddon.php` | 91-92 | `@include_once` of deleted file + `GeoIP::loader()` |
| `wp-slimstat-pro/src/Addon/Addons/MaxMindDetailsAddon.php` | 159 | `GeoIP::get_pack()` — no such method in new API |
| `wp-slimstat/src/Services/Geolocation/GeolocationService.php` | 1-34 | New geolocation API entry point |
| `wp-slimstat/src/Services/Geolocation/Provider/MaxmindGeoIPProvider.php` | 33-69 | New `locate()` returns flat array, not raw MaxMind records |

### Security Note

If implementing a fix based on this analysis:
- Verify the fix doesn't introduce new vulnerabilities
- Run security linting (SAST) on changed code
- Check for OWASP Top 10 in any new code paths

---

*Validated via `qa-issue-validate` skill ([jaan.to](https://jaan.to) plugin)*

**Comment by parhumm** [comment]
# Fix MaxMindDetailsAddon — Replace Removed GeoIP Class with New Geolocation API

## Context

Issue [#182](https://github.com/wp-slimstat/wp-slimstat/issues/182): The free plugin refactored geolocation from `SlimStat\Services\GeoIP` to `SlimStat\Services\Geolocation\*` (commit `7e5ceac8`), but the Pro addon `MaxMindDetailsAddon` was never updated. Fatal "Class not found" error on Advanced Whois AJAX.

## Review Findings Addressed

1. **Version floor**: Runtime version check via `SLIMSTAT_ANALYTICS_VERSION >= 5.4.0` with actionable upgrade message
2. **Provider-aware**: MaxMind/DB-IP supported with appropriate data; Cloudflare explicitly unsupported with explanation
3. **Coupling isolated**: Reader access behind `class_exists` + `\Throwable` guards with `locate()` fallback
4. **DB-missing UX**: Actionable "configure database" message with settings link when DB is missing (not generic error)

## File to Modify

`wp-slimstat-pro/src/Addon/Addons/MaxMindDetailsAddon.php`

## Changes

### 1. Remove import (line 6)

```php
# Before
use SlimStat\Services\GeoIP;

# After — remove entirely, use FQN with guards
```

### 2. Add helper methods (after `forceLocalGeolocationUrl()`, ~line 59)

Four private static methods:

```php
/**
 * Check if the free plugin version supports the new geolocation API.
 */
private static function isCompatibleSlimStatVersion(): bool
{
    return defined('SLIMSTAT_ANALYTICS_VERSION')
        && version_compare(SLIMSTAT_ANALYTICS_VERSION, '5.4.0', '>=');
}

/**
 * Check if geolocation is available AND supports historical IP lookups.
 * Cloudflare only resolves live request headers, not arbitrary IPs.
 */
private static function isWhoisAvailable(): bool
{
    if (!self::isCompatibleSlimStatVersion()) {
        return false;
    }
    $provider = \wp_slimstat::resolve_geolocation_provider();
    return false !== $provider && 'cloudflare' !== $provider;
}

/**
 * Load geolocation data for an IP. MaxMind gets rich data via Reader;
 * DB-IP falls back to GeolocationService::locate() with normalized output.
 *
 * @return array|null Nested array matching MaxMind record structure, or null on failure
 */
private static function loadGeolocationData(string $ip): ?array
{
    $provider = \wp_slimstat::resolve_geolocation_provider();

    // MaxMind: try Reader for rich data (traits, accuracy, metro, full names)
    if ('maxmind' === $provider) {
        $rich = self::loadMaxmindRichRecord($ip);
        if (null !== $rich) {
            return $rich;
        }
        // Reader failed — fall through to locate() API
    }

    // DB-backed providers: use locate() API as fallback
    return self::loadViaLocateApi($ip, $provider);
}

/**
 * Load full MaxMind record via bundled GeoIp2 Reader.
 * Returns raw nested array (jsonSerialize) or null on any failure.
 */
private static function loadMaxmindRichRecord(string $ip): ?array
{
    if (!class_exists('SlimStat\Dependencies\GeoIp2\Database\Reader')
        || !class_exists('SlimStat\Services\Geolocation\GeolocationFactory')) {
        return null;
    }
    try {
        $instance = \SlimStat\Services\Geolocation\GeolocationFactory::create('maxmind');
        $dbPath = $instance->getDbPath();
        if (empty($dbPath) || !file_exists($dbPath)) {
            return null;
        }
        $reader = new \SlimStat\Dependencies\GeoIp2\Database\Reader($dbPath);
        $precision = \wp_slimstat::get_geolocation_precision();
        $record = ('city' === $precision) ? $reader->city($ip) : $reader->country($ip);
        return $record->jsonSerialize();
    } catch (\Throwable $e) {
        return null;
    }
}

/**
 * Load via GeolocationService::locate() and normalize flat result
 * to the nested array format the HTML template expects.
 */
private static function loadViaLocateApi(string $ip, string $provider): ?array
{
    if (!class_exists('SlimStat\Services\Geolocation\GeolocationService')) {
        return null;
    }
    try {
        $precision = \wp_slimstat::get_geolocation_precision();
        $service = new \SlimStat\Services\Geolocation\GeolocationService($provider, ['precision' => $precision]);
        $flat = $service->locate($ip);
        if (null === $flat) {
            return null;
        }
        return [
            'country' => [
                'iso_code' => $flat['country_code'] ?? '',
                'names'    => ['en' => $flat['country_code'] ?? ''],
            ],
            'city' => [
                'names' => ['en' => $flat['city'] ?? ''],
            ],
            'subdivisions' => [
                ['iso_code' => $flat['subdivision'] ?? ''],
            ],
            'location' => [
                'latitude'        => $flat['latitude'] ?? null,
                'longitude'       => $flat['longitude'] ?? null,
                'metro_code'      => null,
                'accuracy_radius' => null,
            ],
            'postal' => [
                'code' => $flat['postal_code'] ?? '',
            ],
            'traits' => [],
        ];
    } catch (\Throwable $e) {
        return null;
    }
}
```

### 3. Update `forceLocalGeolocationUrl()` (line 53) — prevent dead popup

```php
# Before
if (isset(\wp_slimstat::$settings['addon_maxmind_enable']) && \wp_slimstat::$settings['addon_maxmind_enable'] === 'on') {

# After
if (isset(\wp_slimstat::$settings['addon_maxmind_enable'])
    && \wp_slimstat::$settings['addon_maxmind_enable'] === 'on'
    && self::isWhoisAvailable()) {
```

### 4. Replace guard + loader in `ip2locationIframeContent()` (lines 86-92)

Replace lines 86-92 with three explicit guards + loader:

```php
// Guard 1: Version floor
if (!self::isCompatibleSlimStatVersion()) {
    echo esc_html__('Advanced Whois requires WP SlimStat 5.4.0 or later. Please update the free plugin.', 'wp-slimstat-pro');
    exit(-1);
}

// Guard 2: Geolocation disabled
$provider = \wp_slimstat::resolve_geolocation_provider();
if (false === $provider) {
    echo sprintf(
        __("GeoIP collection is not enabled. Please go to <a href='%s' class='noslimstat'>setting page</a> to enable GeoIP for getting more information and location (country) from the visitor.", 'wp-slimstat-pro'),
        esc_url(\wp_slimstat_admin::$config_url . '2#wp-slimstat-third-party-libraries')
    );
    exit(-1);
}

// Guard 3: Cloudflare cannot look up historical IPs
if ('cloudflare' === $provider) {
    echo esc_html__('Advanced Whois is not available with the Cloudflare geolocation provider. Cloudflare resolves geolocation only for live requests, not historical IP lookups. Please switch to MaxMind or DB-IP for this feature.', 'wp-slimstat-pro');
    exit(-1);
}

// Load geolocation data
$data = self::loadGeolocationData($ip);
if (null === $data) {
    echo sprintf(
        __("The geolocation database is not available. Please go to <a href='%s' class='noslimstat'>setting page</a> to configure and download the geolocation database.", 'wp-slimstat-pro'),
        esc_url(\wp_slimstat_admin::$config_url . '2#wp-slimstat-third-party-libraries')
    );
    exit(-1);
}
```

### 5. Replace get_pack() (line 159)

```php
# Before
<?php if (GeoIP::get_pack() == 'city') { ?>

# After
<?php if (method_exists('\wp_slimstat', 'get_geolocation_precision') && \wp_slimstat::get_geolocation_precision() === 'city') { ?>
```

### 6. Fix object/array notation bug (line 298)

```php
# Before
<?php echo (!empty($data->location->latitude) ? $data->location->latitude : '') . ', ' . (!empty($data->location->longitude) ? $data->location->longitude : ''); ?>

# After
<?php echo (!empty($data['location']['latitude']) ? $data['location']['latitude'] : '0') . ', ' . (!empty($data['location']['longitude']) ? $data['location']['longitude'] : '0'); ?>
```

## Why This Works

| Provider | Data Source | Rich Data? | Notes |
|----------|-----------|------------|-------|
| MaxMind | Reader → `jsonSerialize()` (fallback: `locate()`) | Yes: names, traits, metro, accuracy | Two-tier: Reader first, `locate()` fallback |
| DB-IP | `locate()` → normalized | Basic: code, city, coords | No traits/metro/accuracy; template `!empty()` hides them |
| Cloudflare | N/A | N/A | Explicitly blocked with actionable message |
| Disabled | N/A | N/A | Original actionable message with settings link |
| Old free | N/A | N/A | Version check → "update required" message |

## Critical Files

| File | Role |
|------|------|
| `wp-slimstat-pro/src/Addon/Addons/MaxMindDetailsAddon.php` | Sole file to modify |
| `wp-slimstat/src/Dependencies/GeoIp2/Database/Reader.php` | Rich data source (guarded) |
| `wp-slimstat/src/Services/Geolocation/GeolocationService.php` | Stable `locate()` API fallback |
| `wp-slimstat/src/Services/Geolocation/GeolocationFactory.php` | DB path resolution (guarded) |
| `wp-slimstat/wp-slimstat.php:23` | `SLIMSTAT_ANALYTICS_VERSION` constant |
| `wp-slimstat/wp-slimstat.php:360` | `resolve_geolocation_provider()` |
| `wp-slimstat/wp-slimstat.php:824` | `get_geolocation_precision()` |

## Verification

1. **MaxMind + city DB**: Full whois popup with all fields
2. **MaxMind + country DB**: Country-only popup (city rows hidden)
3. **DB-IP**: Basic whois popup (country code, city, coords)
4. **Cloudflare**: Explicit "not available" message with provider switch guidance
5. **Geolocation disabled**: Actionable "not enabled" + settings link
6. **Missing DB**: Actionable "database not available" + settings link
7. **Old free plugin**: "Requires 5.4.0" update message
8. **Google Maps**: Renders at coords or 0,0 fallback

**Comment by parhumm** [comment]
## E2E Test Results — Fix Verification

PR: https://github.com/wp-slimstat/wp-slimstat-pro/pull/11
Branch: `fix/182-maxmind-details-addon-geoip-migration`

### New tests (6/6 pass)

| Test | Result | Time |
|------|--------|------|
| Admin pages load without PHP fatal errors | :white_check_mark: Pass | 11.2s |
| Whois AJAX responds without fatal error | :white_check_mark: Pass | 2.2s |
| Cloudflare provider blocks whois with explicit message | :white_check_mark: Pass | 6.4s |
| Disabled geolocation shows settings message on AJAX call | :white_check_mark: Pass | 6.2s |
| Admin pages stable when switching providers with addon enabled | :white_check_mark: Pass | 14.6s |
| Addon disabled does not inject whois URL | :white_check_mark: Pass | 4.8s |

### Regression tests (8/8 pass)

All existing geolocation provider pipeline tests pass — no regressions.

### What was tested

1. **No fatal errors** — Admin pages (Dashboard, Reports, Settings) load cleanly with both plugins active across all provider types
2. **AJAX endpoint** — `slimstat_ip2location_iframe_content` returns geolocation HTML or actionable error messages (no PHP fatals)
3. **Cloudflare guard** — Direct AJAX call returns explicit "not available with Cloudflare" message; whois URL not injected into reports page
4. **Disabled guard** — Returns "GeoIP collection is not enabled" with settings link; whois URL not injected
5. **Provider switching** — Cycling through dbip/maxmind/cloudflare/disable with addon enabled causes no errors
6. **Addon toggle** — When addon is off, no whois URL is injected regardless of provider

Spec: `wp-slimstat/tests/e2e/pro-maxmind-details-addon.spec.ts`
